src/TorchSharp/TorchVision/dsets/CIFAR.cs

```diff
@@ -0,0 +1,316 @@
+            {
+#if NETSTANDARD2_0_OR_GREATER
+                var targetDir = NSPath.Join(root, "CIFAR100");
+#else
+                var targetDir = Path.Join(root, "CIFAR100");
+#endif // NETSTANDARD2_0_OR_GREATER
+
+                if (download) Download(targetDir, baseUrl, "cifar-100", "cifar-100-binary.tar.gz");
+
+                this.transform = transform;
+
+                var dataPath = Path.Combine(targetDir, "cifar-100-binary");
+
+                if (train) {
+                    _count = ReadSingleFile(Path.Combine(dataPath, "train.bin"));
+                } else {
+                    _count = ReadSingleFile(Path.Combine(dataPath, "test.bin"));
+                }
+
+            }
+            private int _count = 0;
+
+            private int ReadSingleFile(string path)
+            {
+                const int height = 32;
+                const int width = 32;
+                const int channels = 3;
+
+                byte[] dataBytes = File.ReadAllBytes(path);
+
+                var imgSize = channels * height * width;
+                var recSize = imgSize + 2;
+
+                if (dataBytes.Length % recSize  != 0)
+                    throw new InvalidDataException($"Not a proper CIFAR100 file: {path}");
+
+                int count = dataBytes.Length / recSize;
+
+                // Go through the data and create tensors
+
+                for (var i = 0; i < count; i++) {
+
+                    var recStart = i * recSize;
+                    var imgStart = recStart + 2;
+
+                    // CIFAR100 has two labels -- one, which is 20 categories, one which is 100 entities.
+                    coarse_labels.Add(tensor(dataBytes[recStart], int64));
+                    fine_labels.Add(tensor(dataBytes[recStart+1], int64));
+
+                    var floats = new float[imgSize];
+                    for (int j = 0; j < imgSize; j++) {
+                        floats[j] = dataBytes[j + imgStart] / 256.0f;
+                    }
+
+                    data.Add(tensor(floats, new long[] { channels, width, height }));
+                }
+
+                return count;
+            }
+
+            private List<Tensor> data = new();
+            private List<Tensor> coarse_labels = new();
+            private List<Tensor> fine_labels = new();
+
+            private ITransform transform;
+            public override long Count => _count;
+
+
+            public override void Dispose()
+            {
+                data.ForEach(d => d.Dispose());
+                fine_labels.ForEach(d => d.Dispose());
+                coarse_labels.ForEach(d => d.Dispose());
+            }
+
+            /// <summary>
+            /// Get tensor according to index
+            /// </summary>
+            /// <param name="index">Index for tensor</param>
+            /// <returns>Tensors of index. DataLoader will catenate these tensors into batches.</returns>
+            public override Dictionary<string, Tensor> GetTensor(long index)
```

Comment:
>GetTensor [](http://example.com/codeflow?start=55&length=9)

I am wondering, do we claim this is a thread safe? or we don't care?

Reply:
No, we don't claim that. There's another PR, which is mean to make DataLoader parallel, in which we have to deal with that.